### DISCUSSION TITLE
Could doctrine/lexer v2 be supported?

### DISCUSSION THREAD

**Discussion by lazka** [doctrine/doctrine2#10297] [General]
Currently I get lots of composer errors when installing doctrine/orm after installing symfony, because symfony pulls in doctrine/lexer v2 and doptrine/orm is not compatible and requires a downgrade.

```
    - doctrine/orm[2.12.2, ..., 2.13.4] require doctrine/lexer ^1.2.3 -> found doctrine/lexer[1.2.3] but the package is fixed to 2.0.0 (lock file version) by a partial update and that version does not match. Make sure you list it as an argument for the update command.
    - symfony/orm-pack[v2.2.0, ..., v2.3.0] require doctrine/orm * -> satisfiable by doctrine/orm[2.1.3, ..., 2.13.4].
    - doctrine/orm 2.2.3 requires doctrine/dbal 2.2.* -> satisfiable by doctrine/dbal[2.2.0, 2.2.1, 2.2.2].
    - doctrine/orm[2.2.0, ..., 2.2.2] require doctrine/dbal >=2.2.1,<2.2.99 -> satisfiable by doctrine/dbal[2.2.1, 2.2.2].
    - doctrine/orm[2.1.3, ..., 2.1.7] require doctrine/dbal 2.1.* -> satisfiable by doctrine/dbal[2.1.5, 2.1.6, 2.1.7].
```

**Comment by derrabus** [comment]
Yes, it could.

**Comment by greg0ire** [comment]
I'm working on it, but even if I didn't there shouldn't be any issue anywhere. Maybe you are misusing Composer though? It's hard to tell from just the upgrade you provided.

> the package is fixed to 2.0.0 (lock file version) by a partial update and that version does not match.

Why is it fixed? `composer why doctrine/lexer 2.0.0`?

**Reply by derrabus** [reply]
If you have a project with doctrine/annotations only, Lexer will be locked at v2. If you now try to install the ORM using the `composer require` command, Composer will refuse to downgrade Lexer and thus to install the ORM. The solution is to run `composer require` with `--no-update` and run `composer update` afterwards.

**Reply by lazka** [reply]
yes, sorry, I should have added steps to reproduce the issue:

```console
$ composer require doctrine/annotations
$ composer require doctrine/orm
  Problem 1
    - Root composer.json requires doctrine/orm ^2.13 -> satisfiable by doctrine/orm[2.13.0, ..., 2.13.4].
    - doctrine/orm[2.13.0, ..., 2.13.4] require doctrine/lexer ^1.2.3 -> found doctrine/lexer[1.2.3] but the package is fixed to 2.0.0 (lock file version) by a partial update and that version does not match. Make sure you list it as an argument for the update command.
```

I know this is easy to fix on the user side by allowing dependency up/downgrades. But I got a user reporting a problem because one of my packages depending on doctrine/orm failed to install this way :)

**Reply by greg0ire** [reply]
@lazka out of curiosity, why is your customer using `composer require doctrine/annotations`? That package is deprecated, using it seems like a bad idea when starting a new project.

**Reply by lazka** [reply]
It gets pulled in via `doctrine/doctrine-bundle`. The above example is made up/reduced to keep it simple.

**Reply by derrabus** [reply]
But to be fair, this is a DX issue of Composer, not a Doctrine one. We are in the process of enabling all of our packages for Lexer 2 one package at a time.

tbh, opening issues hours after the Lexer 2 release because Composer complains about a solvable problem, that's a bit rude.

**Reply by greg0ire** [reply]
> It gets pulled in via doctrine/doctrine-bundle. 

Ah yeah indeed `doctrine/doctrine-bundle` should probably not have that requirement :thinking: 


I must admit I agree with @derrabus : were you seriously thinking we would not add support for a new version of another `doctrine/` package? Or were you telling us to hurry up?

**Reply by lazka** [reply]
I'm sorry, I was not aware of the timing there. I was just starring at a very long composer error message trying to figure out the cause and finally ending up here :).

Feel free to ignore my question then.

Thanks for your work.

**Reply by greg0ire** [reply]
No worries, I guess you were just clumsy when picking that title.
